scripts/odbpy/defutil.py

```diff
@@ -52,249 +51,92 @@ def extract_core_dims(output_data, input_lef, input_def):
-            collecting = True
-
-    # assert(len(def_two_out_lines) == def_two_count) # sanity check
-    final_out_lines = []
-
-    def_one_count = 0
-    for line in def_one_lines:
-        start_match = section_start_rx.search(line)
-        end_match = section_end_rx.search(line)
-        if start_match is not None:
-            def_one_count = int(start_match[1])
-            final_count = def_one_count
-            if not replace_two:
-                final_count += def_two_count
-            final_out_lines.append(f"{item} {final_count} ;")
-        elif end_match is not None:
-            if not replace_two:
-                final_out_lines += def_two_out_lines
-            final_out_lines.append(f"END {item}")
-        else:
-            final_out_lines.append(line)
-
-    return "\n".join(final_out_lines)
-
-
 @click.command("merge_components")
-@click.option("-o", "--output", default="./out.def")
-@click.option("-l", "--input-lef", required=True, help="Merged LEF file")
-@click.argument("def_one")
-@click.argument("def_two")
-def merge_components(output, input_lef, def_one, def_two):
-    # TODO: Rewrite in OpenDB if possible
-    def_one_str = open(def_one).read()
-    def_two_str = open(def_two).read()
-
-    with open(output, "w") as f:
-        f.write(merge_item_section("COMPONENTS", def_one_str, def_two_str))
-
-
-cli.add_command(merge_components)
-
-
-@click.command("merge_pins")
-@click.option("-o", "--output", default="./out.def")
-@click.option("-l", "--input-lef", required=True, help="Merged LEF file")
-@click.argument("def_one")
-@click.argument("def_two")
-def merge_pins(output, input_lef, def_one, def_two):
-    # TODO: Rewrite in OpenDB if possible
-    def_one_str = open(def_one).read()
-    def_two_str = open(def_two).read()
-
-    with open(output, "w") as f:
-        f.write(merge_item_section("PINS", def_one_str, def_two_str))
-
-
-@click.command("move_diearea")
-@click.option("-l", "--input-lef", required=True, help="Merged LEF file")
 @click.option(
-    "-o",
-    "--output-def",
+    "-w",
+    "--with-components-from",
+    "donor_def",
     required=True,
-    help="Output DEF. File should exist. Die area will be applied to this DEF",
+    help="A donor def file from which to extract components.",
 )
-@click.option("-i", "--template-def", required=True, help="Input DEF")
-def move_diearea_command(input_lef, output_def, template_def):
+@click_odb
+def merge_components(reader, donor_def, input_lef):
     """
-    Move die area from input def to output def
+    Adds all components in a donor DEF file that do not exist in the (recipient) INPUT_DEF.
+
+    Existing components with the same name will *not* be overwritten.
     """
-    move_diearea(input_lef, output_def, template_def)
+    donor = OdbReader(input_lef, donor_def)
+    recipient = reader
+
+    for instance in donor.instances:
+        odb.dbInst_create(recipient.block, instance.getMaster(), instance.getName())
```

Comment:
will this also place the instance to the same position in the donor def?

Reply:
No, that's not the intention.